### PR TITLE
Refactor settings

### PR DESCRIPTION
**Pull request by thbonk**
Refactored the entire library to omit the Settings singleton. This is to enable multi-server/-user clients. The changes are massive and are not compatible with earlier versions of the framework.

### CODE REVIEW

File: Sources/Targets/Accounts.swift [hunk 1]
```diff
@@ -3,54 +3,61 @@ import Moya
...
+        case relationships(URL, AccountId)
+        case search(URL, SearchQuery, Int)
     }
 }
 
 extension Mastodon.Account: TargetType {
     public var baseURL: URL {
-        return Settings.shared.baseURL!.appendingPathComponent("/api/v1/accounts")
+        switch self {
+        case .account(let url, _), .verifyCredentials(let url), .followers(let url, _),
```
Comment (DK1DA) [inline Sources/Targets/Accounts.swift:25]: Can't we use use `default:` here?

Comment (DK1DA) [reply Sources/Targets/Accounts.swift:25]: Ah right, we can't because of the parameter...

**Review by kimar — COMMENTED** [review]
TBH I'm not convinced this is the optimal way to go, it seems like this will add a lot of boilerplate code, specifically inside the `switch` statements, this rather looks like code-smell to me we shouldn't add because it likely won't scale very well (once we're adding more APIs or modifying existing ones).

I agree however that we need a better solution because working with the singleton also won't scale in terms of multi-tenancy. I'd like to have a look how others are solving this using Moya...

**Comment by thbonk** [comment]
IMO we need to make sure that there is no central point where the URL is coming from. Imagine this use case:
- one account already exists and is active
- user wants to login to another account on another server
- for application registration and OAUTH purposes, you need to trigger requests using another base URL
- switching this base URL in a central place error prone

I'm don't know much about Moya so this was my best guess for now. Looking forward to a better proposal :-)

**Comment by DK1DA** [comment]
I agree, I'm also not a 100% about the best place for this in Moya, I've found this:

I think what we want to achieve is to actually set the endpoint in the `endpointClosure`: https://github.com/Moya/Moya/blob/master/docs/Endpoints.md this way we don't have to bloat the enums. Just had a quick glimpse at it, not sure if that's actually the right place.

**Comment by thbonk** [comment]
I'm going to do other changes based on the endpoint approach. I keep you updated.

**Comment by DK1DA** [comment]
Inwould ❤️ to contribute/help but unfortunately I won't have a chance to really contribute a lot in the next couple of weeks 😢 but I'll try to push some code wherever I can

**Comment by thbonk** [comment]
These changes don't break the compatability, and I'm almost done with them. I keep you updated.

**Pull request closed without merging**